wazuh/wazuh
# Report non-applicable checks for SCA

This is the PR for the issue https://github.com/wazuh/wazuh/issues/3029

Two new fields have been added. This occurs when a check cannot be run because the file or directory do not exist, an execution of command has returned timeout, cannot open a registry, etc.

For example, the next check could not be run because there is no field named inted.conf in my specific system:
```
 - id: 5030
   title: "Ensure telnet client is not installed"
   description: "The telnet package contains the telnet client, which allows users to start connections to other systems via the telnet protocol."
   rationale: "The telnet protocol is insecure and unencrypted. The use of an unencrypted transmission medium could allow an unauthorized user to steal credentials. The ssh package provides an encrypted session and stronger security and is included in most Linux distributions."
   remediation: "Run the following command to uninstall telnet: # apt-get remove telnet"
   compliance:
    - cis_csc: "2.6, 4.5"
    - cis: "2.3.4"
    - pci_dss: "2.2.3"
   condition: any
   rules:
     - 'f:/etc/inetd.conf -> !r:^# && r:telnet;'
```
This is its entry at the database:

```
1017679814|5030|cis_debian|Ensure telnet client is not installed|The telnet package contains the telnet client, which allows users to start connections to other systems via the telnet protocol.|The telnet protocol is insecure and unencrypted. The use of an unencrypted transmission medium could allow an unauthorized user to steal credentials. The ssh package provides an encrypted session and stronger security and is included in most Linux distributions.|Run the following command to uninstall telnet: # apt-get remove telnet||||||||Not applicable|File /etc/inetd.conf not found.
```

And the alert generated:

```
** Alert 1555491173.684284: - sca,gdpr_IV_35.7.d
2019 Apr 17 10:52:53 misi->sca
Rule: 19009 (level 3) -> 'CIS benchmark for Debian/Linux: Ensure telnet client is not installed. Not applicable. Reason: File /etc/inetd.conf not found.'
{"type":"check","id":249621373,"policy":"CIS benchmark for Debian/Linux","policy_id":"cis_debian","check":{"id":5030,"title":"Ensure telnet client is not installed","description":"The telnet package contains the telnet client, which allows users to start connections to other systems via the telnet protocol.","rationale":"The telnet protocol is insecure and unencrypted. The use of an unencrypted transmission medium could allow an unauthorized user to steal credentials. The ssh package provides an encrypted session and stronger security and is included in most Linux distributions.","remediation":"Run the following command to uninstall telnet: # apt-get remove telnet","compliance":{"cis_csc":"2.6, 4.5","cis":"2.3.4","pci_dss":"2.2.3"},"file":"","status":"Not applicable","reason":"File /etc/inetd.conf not found."}}
sca.type: check
sca.scan_id: 249621373
sca.policy: CIS benchmark for Debian/Linux
sca.check.id: 5030
sca.check.title: Ensure telnet client is not installed
sca.check.description: The telnet package contains the telnet client, which allows users to start connections to other systems via the telnet protocol.
sca.check.rationale: The telnet protocol is insecure and unencrypted. The use of an unencrypted transmission medium could allow an unauthorized user to steal credentials. The ssh package provides an encrypted session and stronger security and is included in most Linux distributions.
sca.check.remediation: Run the following command to uninstall telnet: # apt-get remove telnet
sca.check.compliance.cis_csc: 2.6, 4.5
sca.check.compliance.cis: 2.3.4
sca.check.compliance.pci_dss: 2.2.3
sca.check.status: Not applicable
sca.check.reason: File /etc/inetd.conf not found.
```

These checks will return a status field which states that the check is 'not applicable' and the reason why. If a check has two or more rules, in case that one of them can't be run, the whole check will be invalid returning the reason from that specific rule. In case there is more than one invalid rule, the check will return the reason of the last one.

Testing
---------
- Valgrind `analysisd`.
- Valgrind `modulesd`.
- Having a check that has a rule with the name of a file that doesn't exist returns a `passed` status.
- Having a check that has a rule with the name of a file that doesn't exist and has pattern returns a `not applicable` status with its specific reason.
- Having a check that has a rule with the name of a directory that doesn't exist returns a `passed` status.
- Having a check that has a rule with the name of a directory that doesn't exist and has file pattern returns a `not applicable` status with its specific reason.
- Running a command at the agent side with the option `remote_commands` disabled returns a `non applicable` check with its specific reason.
- The policy file summary has two new fields, `invalid` which states the number of invalid checks run and `total_checks`, which is the total number of checks passed.
- 
- Check that an alert is generated when a new invalid check appears.
- Check that alerts are generated when a previous check changes its status from `passed` or `failed` to `not applicable`.
- Check that the DB is filled when running a non-applicable check.
- Check that the DB is updated when a check changes its status to `not applicable`.
